### DISCUSSION TITLE
V Language Review (2022)

### DISCUSSION THREAD

**Discussion by vinicius91carvalho** [vlang/v#15614] [General]
Are the concerns valid?

https://mawfig.github.io/2022/06/18/v-lang-in-2022.html

**Comment by medvednikov** [comment]
No, this review is about a couple of simple type checker bugs, which have been fixed. You can try it yourself.

Since this was an anonymous attack on the language, this review won't get updated of course.

**Comment by Wajinn** [comment]
> Are the concerns valid?

No, they aren't and neither is the blog.  That was a "hit and run" type of attack on the language.  The primary purpose was to create a negative blog to spread negative propaganda and try very hard to convince others to not use it, as if they need to be told what to think.  They then linked to it from other websites, using their troll accounts.  It attempts to create the disguise of seeming like it's somewhat legitimate to those that don't know any better or don't know about Vlang's dealings with advocates of competing languages.  Notice that the creator of the blog has disappeared, and that his target was very specifically Vlang.  That should be the first clue, as to what is really going on.

As the creator of the attack piece has a GitHub account, they could have very easily posted their opinions on here too.  Either in discussions or issues, or both.  Just like you (the OP) did.  If they were really concerned or interested in the language, they could have addressed the Vlang GitHub community directly.  If they were as technically competent, as the impression they were trying to give, then they can debate the merits of their opinions with other advanced programmers and developers.

There are those that feel the success of Vlang is a threat to their interests, so apparently feel compelled to constantly attack it or sneakily instigate others to troll for them.  They jump to various websites and blogs to launch unnecessary and unprovoked attacks or attempt to convert others to do so for them.  This behavior doesn't make any sense, if they are so satisfied with the language they are using or are not using and not interested in Vlang (as they might claim).  Yet, very oddly, they feel compelled to keep mentioning and checking on a language that they have never used, don't use, or are using something else.  The reality is they are quite worried about how well Vlang is doing and progressing, which is a credit to the hard work of its contributors and developers.

**Reply by ShalokShalom** [reply]
I can see that some people might feel threatened by V for various issues. Why do you think this is the case?

**Reply by Wajinn** [reply]
> I can see that some people might feel threatened by V for various issues. Why do you think this is the case?

Excuse the delay in response.  From studying this, there appears to be strong competitiveness for support, funding, popularity, and rankings among the newer programming languages.  V doing well can cause some to irrationally feel resentment (support or funding), insecure (about keeping a popularity lead or language ranking), or that their interests are threatened.  

This has then bubbled over into some getting into uncalled for acts of unscrupulousness and trolling, in sneaky attempts to create negative public opinions about rivals.  This is how we get ridiculous lies such as V is fake, vaporware, doesn't work, etc...

The blog in question, is another example of that, where the person who created it (using throwaway troll accounts) went on a "media blitz" of pushing their negative propaganda on multiple websites.  It was not at all "innocent", but rather they were trying to inflict as much reputational damage as they could.  You then can get people that haven't or don't use the language, then spouting back lies and falsehoods, as the result of reading misinformation and instigation by trolls.  However, despite this low and crass behavior, V will persevere and do great things.  Many, when finding out how improved or good V is, will be won over.

**Reply by ShalokShalom** [reply]
But why do I see this behavior just raised against V?

**Reply by Wajinn** [reply]
> But why do I see this behavior just raised against V?

From my observations, that is not the case. There is lots of tribalistic bad behavior among languages. Well known historical examples are what Pascal, Ruby, PHP, Lisp, Smalltalk, and others have had to go through.  Attacks on a language can have unseen interests behind them that are instigating and encouraging the bad behavior or trolling, to include recruiting people through misinformation, for their quiet benefit and thinking it will be a "come up" for them.

My opinion is that V's qualities and potential elicit fears, insecurities, and resentment among its rivals.  Where they feel that V's rise, is their downfall or it got/is getting the support and attention they felt/feel their language should have.  Which is why they can be oddly obsessed with it.  If they are so satisfied with whatever they are using or doing, they should not be concerned or worried about V, yet they very much are.

**Reply by dumblob** [reply]
> But why do I see this behavior just raised against V?

That is certainly an important and valid question.

But as @Wajinn said, it is probably a bubble around us - in your bubble there might be less of threatening against other languages than in other people's bubbles.

Still we must not pretend the historical "public communication" of *V* was OK because it was **not**. And I for myself am convinced this is the building block for spreading misinformation - you make a nasty mistake once and even if it is already fixed, black hats will try to bring this up over and over again (in new forms/reincarnations).

[First they ignore you, then they laugh at you, then they fight you, then you win!](https://en.wikipedia.org/wiki/Nicholas_Klein#Address_to_the_Clothing_Workers )

**Reply by ShalokShalom** [reply]
Yeah, I also think it was a combination of bad communication and bad intent...

**Comment by axetroy** [comment]
I've read a lot of these articles over the past few years.

In my opinion, the culprit is the high expectations given to people at the beginning of V's release.

it gives people a feeling

1. V is a language that combines Rust and Go, simple, safe, and fast.
2. V It's fast, it's safe, it's a system-level language
3. V has many functions, can write games, can write cross-platform UI, can write an operating system, can write Web services
...

People look at it with such high expectations and found out that the HTTP module actually uses the curl command, and it all looks unreal.

so, Don't advertise too early that hasn't been implemented yet

**Reply by JalonSolov** [reply]
Where in the world do you see the use of `curl` in the http module?  I just searched through all the V source, and the only place curl is use is in github actions, not the http module.

The word `curl` does appear in the list of MIME types, but the `curl` command is _not_ used.

Please be sure before spreading misinformation.

**Reply by larpon** [reply]
If my memory serves me well - the http module *did* use curl in a system call several years ago just after the public release - IIRC it was fixed relatively fast when the net module matured. You can [probably find it in the git log](https://github.com/vlang/v/blob/d32e538073e55c603992b5b65ebc837b01c28576/http/download_lin.v#L55) if you dig deep enough.

**Reply by JalonSolov** [reply]
Sad, but the implication of what I read is that it _still_ uses curl, and that is simply untrue.

**Reply by larpon** [reply]
Yes, ofc - no curl in use in V today 🙂 (...and hasn't been in use for several years IIRC)

**Reply by medvednikov** [reply]
V doesn't use curl in the http module, don't spread misinformation.

**Reply by ShalokShalom** [reply]
Why was it used?

**Reply by dumblob** [reply]
@ShalokShalom as a temporary solution to implement and showcase the API. As the quote goes [*Give me an API, and I shall move the earth*](https://en.wikiquote.org/wiki/Archimedes ).

**Comment by Wajinn** [comment]
> People look at it with such high expectations and found out that the HTTP module actually uses the curl command, and it all looks unreal.

The misinformation that you have unfortunately regurgitated, is part of the bizarre world of attacks on the V language that people are not understanding or are clearly aware of. People should be aware that V has certain competitors and detractors who will unscrupulously spread falsehoods, troll, or unfairly criticize a young and still developing language. This includes being behind the scenes to convert and push others to troll for them, attack, and spread the false claims and misinformation they are feeding them.

We are dealing with certain competitors that are of the mindset that being so unscrupulous and sneaky is beneficial to them. They have an agenda to stop or impede V from becoming more popular. They can be under the belief this will help their language become more successful, are angry their language is not as successful, or can gain more supporters and backers through their distortions and slander. This is partially why there are these strange blogs and fake reviews that continual post links to old and outdated blog entries, use troll accounts, or are part of troll gatherings for the purposes of bashing a competing language. They are of the hope that these type of unethical tactics and attempts to artificially create bad press can shape public opinion. People will have heard or read bits of something, from somewhere, and then form false beliefs based on their lies without having actually used V or clearly investigating for themselves.

**Reply by ras-drive** [reply]
> V has certain competitors and detractors who will unscrupulously and continuously put out such misinformation, troll, or unfairly criticize a young and still developing language for their own purposes and gains

I just don't see why, programming languages that are free (looking at you MatLab) have no reason to try and purposely sabotage one another. I myself am very skeptical of the language and think the claims are enormous but if you can prove it then more power to you.

**Reply by larpon** [reply]
IMO no language, or product in general, needs to prove anything.

Anyone can and should simply judge the language (or product) by themselves.
There is no other way to be certain about anything than by testing it yourself.

You can apply this to most things in life that are subjective:
* To know if X song is good - you have to listen to it.
* To know if X clothing looks good on you - you need try it on first.
* To know if X soft drink tastes good - you have to drink it first.

It would save so much useless debate (and advertising!) if everyone would just, please, judge things by themselves - instead of blindly relying on other people to tell them what *they* think - this process is called *advertising* and many people hate it - yet many people seem to do too little to try and eradicate it. The world is an odd place sometimes.
 
Try an install it, see if you like it - move on if you don't - use it if you do - stay and help if you find something could be improved. Simple.

No need for wall-of-text debates and wasting other people's time - it's counterproductive for all parts involved.

**Reply by Wajinn** [reply]
> People will have heard or read bits of something, from somewhere, and then form a belief based on that without having actually used V or clearly investigating.

> Anyone can and should simply judge the language (or product) by themselves. There is no other way to be certain about anything than by testing it yourself.

Totally agree.  Instead of having others tell you what you should think, its usually better to: 1) See for yourself. 2) Make sure that whatever is right for you. 3) At least reserve judgment.  These options are usually better than to have heard some rumors or misinformation from somewhere, and then to regurgitate as if the gospel truth.  

For example, we have people thinking V's HTTP module uses the curl command and other things that are false.  Thus demonstrating they aren't using V or have verified the validity of statements made by detractors (who can have an agenda for distributing misinformation).

At the end of day, if a person is to use some language or product, they should try it for themselves.  If it is not for them, OK, we move on or use whatever is working for us.  No need to engage in theatrics, hostility, or trolling.  If for example, a product falls short of their expectations, then they can make a report to the maker of it (GitHub discussions and issues works quite well).  If what they are saying is valid, good manufacturers or producers can take that into account for improvements to be made.  Then the next version of what they make becomes better.  A win/win for everyone.  

If whatever is being tried is for them or improvements now make it as useful as needed, great, they have found a new treasure that provides benefit.  Maybe even donate in some way or thank the creators of it.  No troll blogs, hate posts, continual theatrics, etc... are necessary in the process.  Rather, we can focus on more useful exchanges or being more productive.

**Reply by danawoodman** [reply]
While I agree in theory, all external marketing in a project should match reality and it's unrealistic for a developer to evaluate every claim in external communications. Sure, it's ultimately up to them to verify the things they care about, but the claims should hold up to scrutiny. I'm not saying V doesn't hold up, just that the argument of "you just need to try it" is a somewhat unrealistic expectation for someone to try out every single claim a software product makes.

**Reply by ras-drive** [reply]
I fully believe that if there is no one showing their opinion on a language many won't even try it. I basically only code in Rust these days and I fully admit that it was because of people talking about how good it is, but after using it for almost a year I know the language pretty damn well and love it!

**Reply by larpon** [reply]
> the argument of "you just need to try it" is a somewhat unrealistic expectation for someone to try out every single claim a software product makes.

Agreed - you can't test *everything* (I think that's more or less the job for a CI server, for which V has one of the most [comprehensive test-suites I've ever seen](https://github.com/vlang/v/tree/master/.github/workflows)... it's **huge** ).
What you *can* test, though, is the language against your own kind of code/projects. 

Doing just that and viewing the results is really what matters IMO.

For me, personally, V shines through every other language that I know and has sky rocketed my productivity. I just can't deny that fact - even if someone comes along and tell me the opposite :man_shrugging: - for me the language works so that's how I'm judging it. Others will have to try it for themselves, throwing their own use-cases at it and see how it goes :slightly_smiling_face:

**Reply by Wajinn** [reply]
> I basically only code in Rust these days...

Let's not pretend that when Rust fans see people trolling and spreading false information about their language, that they don't get upset by it.  Furthermore, V is a much younger language that wasn't hatched from a major company, so is less shielded and is going to have a more difficult road to navigate through shenanigans.

> I fully believe that if there is no one showing their opinion on a language many won't even try it. 

This is true, and I agree with you about that.  But, actual informed opinions or at least useful demonstrations is a different thing than purposefully trolling and spreading misinformation.  From my observations, people in the V community aren't opposed to fair or valid criticism.  That's healthy, and that's how we learn and improve.  To include it's so easy to come over and discuss any particular valid issues at V's GitHub or discord.  Where on the other hand, perpetually spreading lies around the Internet and spewing vitriol doesn't help nor benefit the general public, except for a few that have very specific types of negative agendas.

> V shines... ...for me the language works so that's how I'm judging it.

I agree that V is a great language, and my opinion was also formed after using it.  I very much like both the concept and how it's progressing.

**Reply by Wajinn** [reply]
> I just don't see why, programming languages that are free (looking at you MatLab) have no reason to try and purposely sabotage one another. 

I respectfully disagree. There are organizations (including non-profit types), companies, and individual programmers that get money or receive funding directly (donations, sponsorship, jobs...) or indirectly (books, teaching, videos...) from a language, even though it's free. This can be motivation enough to unethically attempt to sabotage and spread disinformation about other programming languages or developers in their category.

It is maybe naive or disingenuous to think such very targeted continual attacks and distorted language reviews (be it (2022), (2023), or whatever) are not for the purpose of sabotage or to negatively paint an entire community and open-source effort. The GitHub accounts created have obviously only a single purpose, which are being dedicated to smearing V or its developers, while pretending to be a "review". More damning (in general) is when so-called reviewers have backgrounds with clear financial ties to particular competing languages or are using accounts that do not "review" any other languages. This attack is then oddly amplified, by various accounts tied to rival languages and organizations, on various social media platforms. This has even gone so far that detractors will use their control of a social media platform or their greater numbers at a place to ban accounts with positive things to say about V, while allowing its detractors and obvious rivals to troll, spam, misinform, and insult (even if such is against the site's policies).

To more clearly see the point, contrast the characteristics I've described with a real 3rd party [review of V](https://www.youtube.com/watch?v=dmVKerNY-fQ) by Mike Shah. 1) He has reviewed many programming languages, not just singled out or targeted V. 2) He does not engage in provocative trolling or attacks on V, its developers, or its community. 3) He is not tied to the V organization in any way. 4) Is not working for or paid by any language organization for his reviews or opinions (AFAIK). 5) Has actual and very clear credentials to be discussing programming, a Ph.D and years of experience on the subject.

A problem is that various people don't see or are pretending to not know the difference. Mike Shah really evaluates and reviews different programming languages, versus various others who are engaged in attempts at sabotage and attacks, which are disguised as a "review" or "evaluation". Others who have made real reviews of V, are people like: Mihalis Tsoukalos (Linux Format Magazine), Erdet Nasufi (DebConf and published papers on ieee), Massimo Gismondi (Hacker Public Radio), Hazem Abbas (MEDevel), Kevin Da Silva (Medium), and more. These are more neutral 3rd parties with clear credentials (not linked to language creators of competing languages or anonymous accounts) and are identifiable persons, who have discussed and reviewed multiple other programming languages, in addition to V.

**Comment by cn-cq-doudou** [comment]
Hello, I have been paying attention to V for some time. I am concerned about when a mature IDEA plug-in will come out. Currently, the plug-in on VScode only has the function of syntax highlighting

**Reply by JalonSolov** [reply]
The vscode plugin does a lot more than that, when used with the V language server.  There are still issues, though.

The plan has always been to make the vscode plugin work completely, then work on plugins for other editors/ides (including IDEA).

**Reply by JalonSolov** [reply]
@cn-cq-doudou ... and now we have the first public Beta of an IDEA plugin from someone who works for JetBrains!

https://intellij-v.github.io/

**Comment by m4gh3** [comment]


**Reply by larpon** [reply]
Please [install V from source](https://github.com/vlang/v#installing-v-from-source) and try again

**Reply by m4gh3** [reply]
Same results : )
Cloned
added the v repo directory to `PATH` by using `export PATH=[path to V repo]:$PATH`
changed the directory to the ved repo directory
and run the same command as before

**Reply by larpon** [reply]
`valgrind` and V's current garbage collector library doesn't play nice together, since they use some of the same "tricks" to control/intercept/introspect memory calls.

**Reply by dumblob** [reply]
@Larpon that should not be the case with `-autofree`. Is `-autofree` the default now? If not, could you @m4gh3 run it again under valgrind but this time with `-autofree`?

**Reply by larpon** [reply]
@dumblob no `-autofree` is not default.

**Reply by m4gh3** [reply]
Shouldn't have deleted the post but now that I think about it memory leaks are a bit in a gray area when talking about memory safety.
It is also true that Valgrind and Bohem GC do not play well with each other.
Also, no, compiling v with `-autofree` does not seem to improve the results.
